### PR TITLE
Simulate motor noise in the gyros and accels

### PR DESCRIPTION
**Pull request by andyp1per**
This PR enables the simulation of motor-based noise for multicopters. Noise is created at the motor rotational frequency for each motor and combined. The rotational frequency is estimated from the thrust. 

The benefit of this simulation is that it allows filtering targeting motor peaks to be tested in SITL. The support includes a test for the notch filter including dynamic FFT analysis using the batch sampler.

The simulation is controlled by setting 
SIM_VIB_MOT_MAX to the maximum frequency expected from the motors at maximum throttle. The amount of noise added is controlled by the regular SIM_GYR_RND and SIM_ACCEL_RND settings.

### CODE REVIEW

**Comment by andyp1per** [comment]
@peterbarker I would be grateful for a review!

File: Tools/autotest/arducopter.py [hunk 1]
```diff
@@ -3832,8 +3894,11 @@ def tests(self):
              "Log download",
              lambda: self.log_download(
                  self.buildlogs_path("ArduCopter-log.bin"),
-                 upload_logs=len(self.fail_list) > 0))
+                 upload_logs=len(self.fail_list) > 0)),
 
+            ("MotorVibration",
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py]: Needs to go before log download, not after

File: Tools/autotest/common.py [hunk 2]
```diff
@@ -604,7 +605,7 @@ def log_download(self, filename, timeout=360, upload_logs=False):
         """Download latest log."""
         self.wait_heartbeat()
         self.mavproxy.send("module load log\n")
-        self.mavproxy.expect("Loaded module log")
+        self.mavproxy.expect(["Loaded module log", "module log already loaded"])
```
Comment (peterbarker) [inline Tools/autotest/common.py]: This should never be the case.

Why was this required?

Whoever loaded it should unload it :-)

File: Tools/autotest/arducopter.py [hunk 3]
```diff
@@ -1448,6 +1449,67 @@ def fly_motor_fail(self, fail_servo=0, fail_mul=0.0, holdtime=30):
...
+
+        hover_time = 15
+        tstart = self.get_sim_time()
+        self.progress("Hovering for %u seconds" % hover_time)
+        while self.get_sim_time_cached() < tstart + hover_time:
+             attitude = self.mav.recv_match(type='ATTITUDE', blocking=True)
+        tend = self.get_sim_time()
+
+        self.do_RTL()
+        vibe_log = self.buildlogs_path("ArduCopter-vibration-log.bin")
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py]: There is a method to get the currently used binary log file.  Any reason not to use that?

**Comment by tridge** [comment]
I'm happy with the SITL C++ changes, just needs peters review on autotest changes

File: Tools/autotest/arducopter.py [hunk 4]
```diff
@@ -1448,6 +1449,67 @@ def fly_motor_fail(self, fail_servo=0, fail_mul=0.0, holdtime=30):
...
+            raise NotAchievedException("Did not detect a motor peak, found %f at %f dB" % (freq, numpy.amax(psd["X"][ignore_bins:])))
+        else:
+            self.progress("Detected motor peak at %fHz" % freq)
+
+        # now add a notch and check that the peak is squashed
+        self.set_parameter("INS_NOTCH_ENABLE", 1)
+        self.set_parameter("INS_NOTCH_FREQ", freq)
+        self.set_parameter("INS_NOTCH_ATT", 50)
+        self.set_parameter("INS_NOTCH_BW", freq/2)
+        self.reboot_sitl()
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py]: Need a try/catch block here as we don't automatically reboot SITL as part of the suite.  If the test dies we may not recover correctly.

File: Tools/autotest/arducopter.py (hunk 3 above)
Comment (andyp1per) [reply Tools/autotest/arducopter.py]: 👍 I did not know about it 😄

File: Tools/autotest/arducopter.py (hunk 4 above)
Comment (andyp1per) [reply Tools/autotest/arducopter.py]: Ok

File: Tools/autotest/arducopter.py (hunk 1 above)
Comment (andyp1per) [reply Tools/autotest/arducopter.py]: ok

File: Tools/autotest/common.py (hunk 2 above)
Comment (andyp1per) [reply Tools/autotest/common.py]: Probably just debug. I'll take it out.

File: Tools/autotest/arducopter.py (hunk 3 above)
Comment (andyp1per) [reply Tools/autotest/arducopter.py]: @peterbarker I'm probably just being dense but can you point me at it?

Comment (andyp1per) [reply Tools/autotest/arducopter.py]: I tried current_onboard_log_filepath() but it fails, just going to stick with the current approach

File: Tools/autotest/common.py (hunk 2 above)
Comment (andyp1per) [reply Tools/autotest/common.py]: No, it's required. Not sure who isn't unloading the log module - but it isn't me!

**Comment by peterbarker** [comment]
On Tue, 3 Dec 2019, Andy Piper wrote:

> +        vibe_log = self.buildlogs_path("ArduCopter-vibration-log.bin")
> 
> I tried current_onboard_log_filepath() but it fails, just going to stick with the current approach

That's fine.  What was the failure, 'though?

**Comment by andyp1per** [comment]
> On Tue, 3 Dec 2019, Andy Piper wrote: + vibe_log = self.buildlogs_path("ArduCopter-vibration-log.bin") I tried current_onboard_log_filepath() but it fails, just going to stick with the current approach
> That's fine. What was the failure, 'though?

Something about index out of range

**Comment by andyp1per** [comment]
Thanks for the logging fix @peterbarker - works well!

**Comment by andyp1per** [comment]
@peterbarker any idea why I would get this:

   self.check_rc_defaults()
```
  File "/home/travis/build/ArduPilot/ardupilot/Tools/autotest/common.py", line 967, in check_rc_defaults

    raise NotAchievedException("No RC_CHANNELS messages?!")

NotAchievedException: No RC_CHANNELS messages?!
```

**Comment by andyp1per** [comment]
Ok I have rebased and made some tweaks and is passing now. Once the final change gets blessed by CI I will drop the matek commit and it will be ready for merge.

**Comment by andyp1per** [comment]
@peterbarker I got this failure again:
https://travis-ci.org/ArduPilot/ardupilot/jobs/621076962?utm_medium=notification&utm_source=github_status
I feel like this is a symptom of some deeper problem to do with log downloading. Any ideas? Would cleaning up the logs as we go along make the tests run faster?

**Comment by andyp1per** [comment]
@peterbarker so I just watched this test in progress, there are basically three problems:
* There's only 2min30s left for the test as it comes at the end of a 47min test run and the timeout is 50min - I think we should increase this
* It generates a 17Mb log file per run which takes 40s to download each time - some of this is just going to be because of the batch logging, but I wonder whether we are getting a lot of logging from either side of the test that could be remove. Is there a faster way to get the log?
* The EKF takes forever to settle - is there a way to avoid this or speed this up?

**Comment by peterbarker** [comment]
> @peterbarker so I just watched this test in progress, there are basically three problems:
> 
>     * There's only 2min30s left for the test as it comes at the end of a 47min test run and the timeout is 50min - I think we should increase this

Yep, it's close!  Is it actually running overtime?

>     * It generates a 17Mb log file per run which takes 40s to download each time (there are two passes to the test so 1min20s) - some of this is just going to be because of the batch logging, but I wonder whether we are getting a lot of logging from either side of the test that could be removed. Is there a faster way to get the log?

Rebooting the vehicle (which you are doing) must start a new log, so it's almost certainly All Your Fault :-P

I did suggest the faster way to get the log - reference it directly using `dfreader_for_current_onboard_log`.  That didn't work straight-up - I really only added that to solve my particular need at the time and didn't extensively test it.  Perhaps if you give me instructions on how to reproduce the index-out-of-range bug I could fix that?

>     * The EKF takes forever to settle - is there a way to avoid this or speed this up?

I've been wondering about pushing stream rates way-down and pushing sim-speedup way-up for periods where we don't really care about anything but simple things like this.   Long traverals to waypoints (like the quadplane tests) might also be candidates for this.

But generally, no - there are hard-wired timeouts in the EKF which makes shortening settle-time problematic.

**Comment by andyp1per** [comment]
Ok I have disabled LOG_DISARMED and that has produced a mcuh smaller log that can be downloaded in 5s. Still timing out though ...

**Comment by andyp1per** [comment]
@peterbarker I got the test working with ```dfreader_for_current_onboard_log()``` thanks! Observing the test its slow because of EKF because of GPS fix - is there any way to speed up the GPS fix? I don't remember it being this bad.

**Comment by peterbarker** [comment]
@andyp1per One option might be setting EKF type 10.  I'm really not sure, 'though.

**Comment by andyp1per** [comment]
@peterbarker tests are now clean - good enough? 🙂
